torch/csrc/cuda/nccl.h

```diff
@@ -6,15 +6,52 @@
 #include <c10/cuda/CUDACachingAllocator.h>
 #include <c10/util/Optional.h>
 
-#include <nccl.h>
-
 #include <cstddef>
 #include <vector>
 
 namespace torch {
 namespace cuda {
 namespace nccl {
 
+/* The following are copied from <nccl.h> and redefined in torch::cuda::nccl namespace */
+/* pytorch should only use the following definition within pytorch scope */
+
+
+/* Opaque handle to communicator to ncclComm*, this reinterpret as ncclComm in nccl.cpp */
+typedef void* ncclComm_t;
+
+#define NCCL_UNIQUE_ID_BYTES 128
+typedef struct { char internal[NCCL_UNIQUE_ID_BYTES]; } ncclUniqueId;
+
+/* Error type */
+typedef enum { ncclSuccess                 =  0,
+               ncclUnhandledCudaError      =  1,
+               ncclSystemError             =  2,
+               ncclInternalError           =  3,
+               ncclInvalidArgument         =  4,
+               ncclInvalidUsage            =  5,
+               ncclNumResults              =  6 } ncclResult_t;
+
+/* Reduction operation selector */
+typedef enum { ncclSum        = 0,
+               ncclProd       = 1,
+               ncclMax        = 2,
+               ncclMin        = 3,
+               ncclNumOps     = 4 } ncclRedOp_t;
+
+/* Data types */
+typedef enum { ncclInt8       = 0, ncclChar       = 0,
+               ncclUint8      = 1,
+               ncclInt32      = 2, ncclInt        = 2,
+               ncclUint32     = 3,
+               ncclInt64      = 4,
+               ncclUint64     = 5,
+               ncclFloat16    = 6, ncclHalf       = 6,
+               ncclFloat32    = 7, ncclFloat      = 7,
+               ncclFloat64    = 8, ncclDouble     = 8,
+               ncclNumTypes   = 9 } ncclDataType_t;
+
```

Comment:
Can you use `enum class` instead of C-style enums (and if aliases are not allowed in enum class defined them using `using` construct?
```suggestion
enum class ncclResult { 
               Success                 =  0,
               UnhandledCudaError      =  1,
               SystemError             =  2,
               InternalError           =  3,
               InvalidArgument         =  4,
               InvalidUsage            =  5,
               NumResults              =  6,
                };

/* Reduction operation selector */
enum class ncclRedOpt_t { Sum        = 0,
               Prod       = 1,
               Max        = 2,
               Min        = 3,
               NumOps     = 4 };

/* Data types */
enum class ncclDataType { 
               Int8       = 0, 
               Char       = 0,
               Uint8      = 1,
               Int32      = 2, 
               Uint32     = 3,
               Int64      = 4,
               Uint64     = 5,
               Float16    = 6
               Float32    = 7
               Float64    = 8,
               ncclNumTypes   = 9 
               };
               using , ncclDouble    = ncclDataType:Float64;

```